src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,298 @@
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.InvertIf
+{
+    internal abstract partial class AbstractInvertIfCodeRefactoringProvider
+    {
+        protected enum InvertIfStyle
+        {
+            // swap if and else
+            Normal,
+            // swap subsequent statements and if body
+            SwapIfBodyWithSubsequence,
+            // move subsequent statements to if body
+            MoveSubsequenceToElseBody,
+            // invert and generete else
+            WithElseClause,
+            // invert and generate else, keep if-body empty
+            MoveIfBodyToElseClause,
+            // invert and copy the exit point statement
+            WithSubsequenceExitPoint,
+            // invert and generate return, break, continue
+            WithNearmostJump,
+            // just invert the condition
+            WithNegatedCondition,
+        }
+
+        protected abstract class Analyzer<TIfStatementSyntax> : IAnalyzer
+            where TIfStatementSyntax : SyntaxNode
+        {
+            public async Task ComputeRefactoringsAsync(CodeRefactoringContext context, SyntaxNode ifNode)
+            {
+                var ifStatement = (TIfStatementSyntax)ifNode;
+
+                var textSpan = context.Span;
+                var document = context.Document;
+                var cancellationToken = context.CancellationToken;
+
+                var headerSpan = GetHeaderSpan(ifStatement);
+                if (!headerSpan.IntersectsWith(textSpan))
+                {
+                    return;
+                }
+
+                if (ifStatement.OverlapsHiddenPosition(cancellationToken))
+                {
+                    return;
+                }
+
+                if (!CanInvert(ifStatement))
+                {
+                    return;
+                }
+
+                SyntaxNode subsequenceSingleExitPointOpt = null;
+
+                InvertIfStyle invertIfStyle;
+                if (IsElselessIfStatement(ifStatement))
+                {
+                    invertIfStyle = GetInvertIfStyle(
+                        ifStatement,
+                        await document.GetSemanticModelAsync().ConfigureAwait(false),
+                        ref subsequenceSingleExitPointOpt);
+                }
+                else
+                {
+                    invertIfStyle = InvertIfStyle.Normal;
+                }
+
+                context.RegisterRefactoring(
+                    new MyCodeAction(
+                        GetTitle(),
+                        c => InvertIfAsync(document, ifStatement, invertIfStyle, subsequenceSingleExitPointOpt, c)));
```

Comment:
every time i see subsequence it feels wrong to me.  But i don't know english well enough to know if that's teh case.  should it not just be "subsequent"?

Reply:
me neither =) but I think either "subsequence" or "subsequent statements" works.

Reply:
(resolved)